File: public/exemplos/c/busca_binaria.c

/*
 * Exemplo de Algoritmo - Busca Binária
 * Professor: Vagner Cordeiro
 * Disciplina: ARA0174 - Algoritmos e Complexidade
 */

#include <stdio.h>
#include <stdlib.h>

// Busca binária recursiva
int busca_binaria(int arr[], int esquerda, int direita, int x) {
    if (direita >= esquerda) {
        int meio = esquerda + (direita - esquerda) / 2;
        
        if (arr[meio] == x)
            return meio;
        
        if (arr[meio] > x)
            return busca_binaria(arr, esquerda, meio - 1, x);
        
        return busca_binaria(arr, meio + 1, direita, x);
    }
    return -1;
}

int main() {
    int arr[] = {2, 3, 4, 10, 40};
    int n = sizeof(arr) / sizeof(arr[0]);
    int x = 10;
    
    int resultado = busca_binaria(arr, 0, n - 1, x);
    
    if (resultado == -1)
        printf("Elemento não encontrado\n");
    else
        printf("Elemento encontrado no índice %d\n", resultado);
    
    return 0;
}

File: public/exemplos/python/busca_binaria.py

"""
Exemplo de Algoritmo - Busca Binária
Professor: Vagner Cordeiro
Disciplina: ARA0174 - Algoritmos e Complexidade
"""

def busca_binaria(arr, esquerda, direita, x):
    """Busca binária recursiva - O(log n)"""
    if direita >= esquerda:
        meio = esquerda + (direita - esquerda) // 2
        
        if arr[meio] == x:
            return meio
        elif arr[meio] > x:
            return busca_binaria(arr, esquerda, meio - 1, x)
        else:
            return busca_binaria(arr, meio + 1, direita, x)
    
    return -1

def main():
    arr = [2, 3, 4, 10, 40]
    x = 10
    
    resultado = busca_binaria(arr, 0, len(arr) - 1, x)
    
    if resultado == -1:
        print("Elemento não encontrado")
    else:
        print(f"Elemento encontrado no índice {resultado}")

if __name__ == "__main__":
    main()
